File: Creational/Lab5/C#/RealEstate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection.Emit;
using System.Text;
using System.Threading.Tasks;
using C_.Interfaces;

namespace C_
{
    public abstract class RealEstate : AgencyObject, ICloneable, IRentable, IBuyable, ISellable
    {
        public string Address { get; set; }
        public string Description { get; set; }
        public double Area { get; set; }
        public double Price { get; set; }
        public string Status 
        { 
            get => _status != null ? _status : "Unknown";
            set
            {
                if (value == "New" || value == "Old" || value == "Middle aged")
                    _status = value;
                else
                    _status = "Unknown";
            } 
        }
        public Realtor Realtor{ get; set; }
        
        //Bools
        public bool IsRentable { get; set; } = false;
        public bool IsSellable { get; set; } = false;
        public bool IsBuyable { get; set; } = false;
        public bool Furnished { get; set; } = false;
        public bool Parking { get; set; } = false;
        public bool Balcony { get; set; } = false;
        public bool Elevator { get; set; } = false;
        public bool AirConditioning { get; set; } = false;
        public bool IsUtilitiesIncluded { get; set; } = false;
        
        private string _status;

        public RealEstate(string address, string description, double area, double price, string status):base() { 
            Address = address;
            Description = description;
            Area = area;
            Price = price;
            Status = status;
            // Realtor = new Realtor("John Doe", "8 800 555 35 35", "example@gmail.com", 2);
        }

        public override string ToString(){
            return $"Id: {Id}, Address: {Address},\n Description: {Description},\nArea: {Area}, Price: {Price}\nStatus: {Status}\n{Realtor}\nFurnished: {Furnished}, "+
            $"Parking: {Parking}, Balcony: {Balcony}, Elevator: {Elevator}, Air conditioning: {AirConditioning}, Utilities included: {IsUtilitiesIncluded}\n";
        }

        public abstract void Rent(Customer customer, DateTime startDate, DateTime endDate);
        public abstract void Sell(Customer customer);
        public abstract void Buy(Customer customer);
        public abstract object Clone();
        public abstract RealEstate ShallowCopy();

        public static void PrintAllInfoAboutEstates(IEnumerable<RealEstate>  realEstates){
            foreach (RealEstate realEstate in realEstates){
                Console.WriteLine(realEstate.ToString());
            }
            Console.WriteLine("\n");
        }

    }
}

File: Creational/Lab6/Python/real_estate.py
import copy
from abc import *
from realtor import *
from agency_object import *
from datetime import datetime

class RealEstate(AgencyObject, ABC):
    def __init__(self, address, description, area, price, status):
        super().__init__()
        self.address = address
        self.description = description
        self.area = area
        self.price = price
        self._status = status
        self.realtor = None

        # Bools
        self.is_rentable = False
        self.is_sellable = False
        self.is_buyable = False
        self.furnished = False
        self.parking = False
        self.balcony = False
        self.elevator = False
        self.air_conditioning = False
        self.is_utilities_included = False

    @property
    def status(self):
        return self._status if self._status is not None else "Unknown"

    @status.setter
    def status(self, value):
        if value in ["New", "Old", "Middle aged"]:
            self._status = value
        else:
            self._status = "Unknown"

    def __str__(self):
        return (f"Id: {self.id}, Address: {self.address},\n"
                f"Description: {self.description},\n"
                f"Area: {self.area}, Price: {self.price}\n"
                f"Status: {self.status}\n"
                f"{self.realtor}\n"
                f"Furnished: {self.furnished}, "
                f"Parking: {self.parking}, "
                f"Balcony: {self.balcony}, "
                f"Elevator: {self.elevator}, "
                f"Air conditioning: {self.air_conditioning}, "
                f"Utilities included: {self.is_utilities_included}\n")

    @abstractmethod
    def rent(self, customer, start_date, end_date):
        pass

    @abstractmethod
    def sell(self, customer):
        pass

    @abstractmethod
    def buy(self, customer):
        pass

    @abstractmethod
    def clone(self):
        pass

    def shallow_copy(self):
        return self.__class__(**self.__dict__)

    @staticmethod
    def print_all_info_about_estates(real_estates):
        for real_estate in real_estates:
            print(real_estate)
        print("\n")

    def reset(self):
        self.address = ""
        self.description = ""
        self.area = 0
        self.price = 0
        self.status = ""
        self.realtor = None
        self.is_rentable = False
        self.is_buyable = False
        self.is_sellable = False
        self.furnished = False
        self.parking = False
        self.balcony = False
        self.elevator = False
        self.air_conditioning = False
        self.is_utilities_included = False